Import fast tanh impl from eigen

xiaomengy opened the pull request:
Summary:
Import fast tanh impl from eigen.

This approach is about 30% faster than MKL v?Tanh and ouver 50% faster than previous Vec256::tanh().

Differential Revision: D15632834

colesbury reviewed (commented):
Is this precise for both float and double? For comparison glibc claims precision within 2 ulp and sleef claims precision within 1 ulp. (There''s also a faster Sleef implementation with 3.5 ulp)

You should also test that NaN and +/-inf are properly handled.

xiaomengy committed:
Import fast tanh impl from eigen (#21384)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21384

Import fast tanh impl from eigen.

This approach is about 30% faster than MKL v?Tanh and ouver 50% faster than previous Vec256::tanh().

Differential Revision: D15632834

fbshipit-source-id: b7aa05c32a88c4d24603d4e28ae09389248ed765

BIT-silence commented:
> Is this precise for both float and double? For comparison glibc claims precision within 2 ulp and sleef claims precision within 1 ulp. (There''s also a faster Sleef implementation with 3.5 ulp)
> 
> You should also test that NaN and +/-inf are properly handled.
